[✔] CRUD usuarios / CRUD roles y permisos

Farid-NL opened the pull request:
## Que hace y que no
- [x] Búsqueda por `nombre completo` y `correo`
- [x] Ordenamiento asc y desc: 
  - `nombre completo` y `correo`
- [x] Campos mostrados _(Sugerencias?)_:
  - `nombre completo` y `correo`
- [ ] ~~Filtros~~
- [x] Ajustar los `queries` de la URL
  - `s=algo`, en vez de `search=algo`
  - `p=2`, en vez de `page=2`
  - `d=asc`, en vez de `sortDirection=asc`
- [x] Cualquier otra cosa que noten

## Errores 🐛
De momento no hay ninguno que haya notado! 🥳

FaruNL commented:
Gracias, justo estaba viendo que Spatie ofrece muchas facilidades y sería mejor usarlo para los roles y permisos.

En cuanto tenga tiempo lo reviso 😎

Natali-Rojas commented:
No entiendo los detalles de styleci 😭   PD. **ayuda**

FaruNL commented:
Ya se hizo merge de otro PR (#12) con las mismas funcionalidades. Cierro. Gracias!! 🥳

Farid-NL committed:
feat: Añade crud básico de usuarios

Farid-NL committed:
chore: Instala laravel-permission y configura archivos asociados

Además, se publicó los archivos de configuración: migración y permisos

Farid-NL committed:
feat: Añade seeders, crea y asigna roles y permisos

Farid-NL committed:
feat: Añade rutas asociadas a roles y protegidas por un middleware

Farid-NL committed:
feat: Muestra y busca nombre completo

Farid-NL committed:
feat: Añade inputs de apellidos

Farid-NL committed:
fix: Permite la posibilidad que los apellidos sean <null>

Farid-NL committed:
fix: Busca por nombre completo y correo

Farid-NL committed:
feat: Añade vistas dinámicas en función de los permisos

Para redirigir dinámicamente después del login; fue necesario
sobreescribir el funcionamiento de LoginResponse.php para que
considerara los prefijos de las rutas y los roles.

Farid-NL committed:
style: Validaciones usan la notación de arreglo

Farid-NL committed:
style: Añade type hinting y type declaration

Farid-NL committed:
fix: Mejora validaciones

- nombre y apellidos: 'alpha' y considera espacios
- Eliminado 'string', pues se sobre-entiende

Natali-Rojas commented on app/Http/Livewire/Admin/UserController.php:
```diff
@@ -0,0 +1,189 @@
+    {
+        if ($this->edit) {
+            return [
+                'user.name' => ['required', 'regex:/^[\pL\pM\s]+$/u', 'max:255'],
```
¿El nombre y apellidos aceptara puntos?

Farid-NL replied:
Pues podría ser, puede que venga un instructor extranjero con un punto en su nombre si es el caso sería asi el regex, me parece:

`regex:/^[\pL\pM\s\.]+$/u`

Natali-Rojas commented on app/Http/Livewire/Admin/UserController.php:
```diff
@@ -0,0 +1,189 @@
+                'user.name' => ['required', 'regex:/^[\pL\pM\s]+$/u', 'max:255'],
+                'user.apellido_paterno' => ['present', 'regex:/^[\pL\pM\s]+$/u', 'max:255'],
+                'user.apellido_materno' => ['present', 'regex:/^[\pL\pM\s]+$/u', 'max:255'],
+                'user.email' => ['required', 'email', 'max:255', Rule::unique('users', 'email')->ignore($this->user)],
```
Unique, podriamos mejor, ponerlos en las migraciones y checar todas los datos que ser requieran en todas las tablas, opinemos y yo hago los cambios

Farid-NL replied:
Por lo que hay que mostrarle al usuario en donde estuvo el error, es necesaria esta validación.

En producción no se mostrará las excepciones, por lo que un buen mensaje de validación es necesario.

Natali-Rojas commented on app/Http/Livewire/Admin/UserController.php:
```diff
@@ -0,0 +1,189 @@
+                'password' => array_replace($this->passwordRules(), [0 => 'present']),
+                'password_confirmation' => ['present'],
+            ];
+        } else {
```
al crear también se debe verificar que no dupliquen datos no?

Farid-NL replied:
Así es, se verifica la duplicidad del email
```php
'user.email' => ['required', 'email', 'max:255', 'unique:users,email'],
```

Farid-NL replied:
Duda, te referías a la contraseña? duplicidad de contraseña? o como?

Natali-Rojas commented on app/Http/Livewire/Admin/UserController.php:
```diff
@@ -0,0 +1,189 @@
+                    $query->where(DB::raw("REPLACE(CONCAT_WS(' ', name, apellido_paterno, apellido_materno), '  ', ' ')"), 'like', "%$search%")
+                        ->orWhere('email', 'like', "%$search%");
+                })
+                ->when($this->sortField === 'nombre_completo', function ($query) {
```
¿Por qué dos filtros Fa?

Farid-NL committed:
style: Corrige code style

Farid-NL committed:
feat: Añade crud básico de roles

Farid-NL committed:
feat: Añade asignación de permisos a roles

Farid-NL committed:
feat: Muestra los permisos de forma amigable

Se añadió un campo a la tabla de permisos donde se almacena el nombre
'amigable' del permiso.

Farid-NL committed:
refactor: Mejora la obtención de ids de permisos y mejora comentarios

Farid-NL committed:
refactor: Crea permisos y roles de manera más clara y breve

Añade permisos para interactuar con roles

Farid-NL committed:
fix: Mejora separación entre checkbox y hace responsivo el grid

Farid-NL committed:
refactor: Reorganiza la disposición de los inputs

Farid-NL committed:
feat: Asigna roles a los usuarios desde la vista del crud de usuarios

Farid-NL committed:
feat: Muestra los roles en la tabla de crud de usuarios

Farid-NL committed:
style: Elimina método inutilizado

Farid-NL committed:
revert: Revierte 58c0caa21776966bb8061fe17801615e6af097e8

"feat: Añade vistas dinámicas en función de los permisos"

Farid-NL committed:
fix: Corrige nombre del conjunto de permisos 'rol' a 'role'

Farid-NL committed:
feat: Utiliza permisos en vez de roles para acceder a las rutas

Farid-NL committed:
fix: Agrupa rutas por un solo permiso '*.show'

Si el usuario quiere acceder a una ruta con la que solo cuenta con un
permiso listado en los middlewares, no lo dejaría. Pues tendría que
tener sí o sí todos los permisos listados.

Farid-NL committed:
fix: Validación de apellidos acepta valores nulos

refactor: Reinicio de variables agrupadas en método `blankUser()`

Farid-NL committed:
feat: Añade autorización de acciones según permisos en el controlador

Farid-NL committed:
feat: Añade páginas de errores personalizadas

Farid-NL committed:
revert: Elimina middleware de rol (role)

Farid-NL committed:
feat: Permite eliminar solamente a roles diferentes de los bases

Farid-NL committed:
fix: Añade autorización de acciones según permisos en el controlador

Farid-NL committed:
refactor: Añade types, inicializa variables y reorganiza acciones

Farid-NL committed:
refactor: Acorta nombre de controladores por medio de imports

Farid-NL committed:
docs: Añade throws comment

Farid-NL committed:
refactor: Reorganiza el código y solo busca si 'search' está no vacío

Farid-NL committed:
Merge branch 'main' into crud_user

Farid-NL committed:
fix: Corrige nombre de roles y hace validaciones más claras

Farid-NL committed:
fix: Corrige nombre de input

Farid-NL committed:
refactor: Restringe la visibilidad del Super Admin

Además, el Super Admin solo podrá visualizar. No se eliminará ni editará

Farid-NL committed:
feat: Restringe la edición de los roles bases

Farid-NL committed:
feat: Utiliza traits para reutilizar código

RouterUIZ007 reviewed (dismissed):
Creo que ya está  mi reyes... 😁🤪

RouterUIZ007 reviewed (commented):
jejeje me gusto la clasificación de clases, bueno de los traits

Farid-NL committed:
fix: Permite ingresar excepciones de variables que no requieren limpieza

Lo hace mediante el arreglo `$cleanStringsExcept`, siempre y cuando esté
declarado.

Natali-Rojas commented on resources/views/livewire/admin/roles/index.blade.php:
```diff
@@ -0,0 +1,98 @@
+                                        </button>
+                                    @endunless
+
+                                    @unless(in_array($r->name, ['Super admin', 'Administrador', 'Participante', 'Instructor']))
```
Fa aquí estableces que Participante e instructor pueden eliminar?, ¿no deberían no es así?, unicamente el admin o el super admin

Farid-NL replied:
No, no hace eso. Lo que hace es que si el el **rol mostrado en la tabla** resulta ser 'Super admin', 'Administrador', 'Participante' o 'Instructor', no mostrará el botón de eliminar.

Natali-Rojas replied:
Pero aplica solo para los primeros 3 verdad, los creados, todos o solo el admin puede elminar 
![image](https://user-images.githubusercontent.com/36526810/168457674-0547971d-8c7b-4c58-8977-a80abfa67b28.png)
 (imagen con usuario creado)

Farid-NL replied:
Más especificamente: Busca el nombre del rol actual _(que se mostrará como una fila en la tabla, es decir, la iteración actual del `for`)_ dentro de el arreglo `['Super admin', 'Administrador', 'Participante', 'Instructor']`. Si coincide con alguno de estos, pues no muestra el botón de eliminar.

Puedes probarlo entrando con diferentes roles.

Farid-NL replied:
> Pero aplica solo para los primeros 3 verdad, los creados, todos o solo el admin puede elminar

Así es, solo para los roles bases

Natali-Rojas replied:
Si lo probé con los tipos de roles, pero no comprendia porque no se eliminan los primero y los nuevos creados sí, pero viendo y analizando, si es necesario que no se eliminen los de base, pero eso abree algo, espera te hago comentario ahí

Farid-NL replied:
> Si lo probé con los tipos de roles, pero no comprendia porque no se eliminan los primero y los nuevos creados sí, pero viendo y analizando, si es necesario que no se eliminen los de base, pero eso abree algo, espera te hago comentario ahí

Oki doki, espero

Natali-Rojas commented on resources/views/livewire/admin/users/edit_create.blade.php:
```diff
@@ -0,0 +1,74 @@
+                <!-- Rol -->
+                <div class="mt-4 sm:flex-1">
+                    <x-jet-label for="rol" value="Rol"/>
+                    <x-input.select wire:model.defer="role" id="rol" class="mt-1 w-full" name="rol">
```
Algo que debemos de analizar en conjunto es la lógica para la asignación de los roles "instructor y participante", porque no se van a asignar desde la tabla usuarios, ya que tenemos hasta ahora que el rol dependa de la variable "estatus_participante" de la tabla pivote, no que se asigne

Natali-Rojas replied:
A menos que a los usuarios docentes(todos) cuando se cree su usuario se le ponga por defecto "participante", y a los instructores de fuera(de empresas como cisco) se les pongo instructor, pero aún así el rol para participante va a sufrir cambios por el valor de la variable "estatus participante"

Natali-Rojas replied:
Lo pondré en pendiente, pero si se tiene que contemplar para el funcionamiento de que los roles dependan de la variable (que no se hará en una vista)

Farid-NL replied:
Oh ya, creo me quedé con la idea pasada.

Ya que según recordaba, la admin podía cambiar el rol de los participantes (todos) según le sea necesario. Y creo recordar que se había mencionado que sería en la vista de usuarios.

---

Lo que me genera conflicto es justamente ese atributo `estatus_participante`, como se lo asignaremos, pues creo una opción seria que lo rellene a la vez que se asigna el rol de spatie.
